clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+
+    AssignmentReconciler(LogContext logContext,
+                         SubscriptionState subscriptions,
+                         ConsumerMetadata metadata,
+                         BlockingQueue<BackgroundEvent> backgroundEventQueue) {
+        this.log = logContext.logger(getClass());
+        this.subscriptions = subscriptions;
+        this.metadata = metadata;
+        this.backgroundEventQueue = backgroundEventQueue;
+    }
+
+    void startLost() {
+        SortedSet<TopicPartition> partitionsToLose = getPartitionsToLose();
+
+        if (partitionsToLose.isEmpty()) {
+            log.debug("Skipping invocation of {} callbacks as no partitions changed in the new assignment",
+                    ConsumerRebalanceListener.class.getSimpleName());
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        backgroundEventQueue.add(new PartitionLostStartedEvent(partitionsToLose));
+    }
+
+    /**
+     * This method should be invoked to signal the completion of the "{@link TopicPartition lost partition}"
+     * process. Specifically, it is to be executed on background thread <em>after</em> the
+     * {@link ConsumerRebalanceListener#onPartitionsLost(Collection)} callback was executed on the application
+     * thread. It should clear the set of {@link SubscriptionState#assignedPartitions() assigned partitions},
+     * regardless of the set of "lost partitions."
+     *
+     * @param lostPartitions Set of {@link TopicPartition partitions} that were lost
+     * @see AssignmentReconciler
+     */
+    void completeLost(Set<TopicPartition> lostPartitions) {
+        log.debug("{} callbacks were successfully invoked", ConsumerRebalanceListener.class.getSimpleName());
+        subscriptions.assignFromSubscribed(Collections.emptySet());
+    }
+
+    /**
+     * Performs the assignment phase of the reconciliation process as described in the top-level class documentation.
+     *
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     */
+
+    void startReconcile(Assignment assignment) {
+        SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(assignment);
+        SortedSet<TopicPartition> partitionsToAssign = getPartitionsToAssign(assignment);
+
+        if (partitionsToRevoke.isEmpty() && partitionsToAssign.isEmpty()) {
+            log.debug("Skipping invocation of {} callbacks as no partitions changed in the new assignment",
+                    ConsumerRebalanceListener.class.getSimpleName());
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        PartitionReconciliationStartedEvent event = new PartitionReconciliationStartedEvent(
+                partitionsToRevoke,
+                partitionsToAssign
+        );
+        backgroundEventQueue.add(event);
```

Comment:
Using rebalance callbacks is pretty uncommon, I think. Therefore, in most of the case, we won't have one. To be precise, it will be set to `NoOpConsumerRebalanceListener`. In this case, we don't have to go back to the application thread to execute nothing. We should just bypass it here. Have you considered this?

Reply:
Agreed.

See the [other comment](https://github.com/apache/kafka/pull/14357/files#r1362032050).